README.txt
CPSonFive
=========

CPSonFive is a product who contains all the integration between core 
CPS functionality and Five. It's purpose it to provide a place where
this integration can be done so that the integration can be introduced
separately from the CPS products, thereby enabling integration without
requiring you to use the latest development version of CPS.

Currently CPSonFive provides the following integration features:

 - CPS form look for Zope3 schema-based forms.

 - Translatability for form error messages and drop down boxes.

 - A document_navigation widget that opens a CPSNavigation browser window.



CHANGES
Requires:
~~~~~~~~~
-
New features:
~~~~~~~~~~~~~
- Now made for Five 1.3 and later. 
Bug fixes:
~~~~~~~~~~
- The "Updated on" text had stopped working, so I changed it for Five 1.3
  support.
New internal features:
~~~~~~~~~~~~~~~~~~~~~~
-


form/widget.py
# -*- coding: ISO-8859-15 -*-
# (C) Copyright 2005 Nuxeo SARL <http://nuxeo.com>
# Author: Lennart Regebro <regebro@nuxeo.com>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License version 2 as published
# by the Free Software Foundation.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
# 02111-1307, USA.
#
# $Id: calendar.py 24398 2005-06-24 12:34:57Z lregebro $

from cgi import escape
from zope.app.form.browser import TextWidget
from zope.app.form.browser.widget import renderElement
from zope.app.form.browser.exception import WidgetInputErrorView
from zope.app.form.interfaces import ConversionError, WidgetInputError

from Products.CMFCore.utils import getToolByName
try:
    from zope.i18nmessageid import MessageFactory
except ImportError:
    from zope.i18nmessageid import MessageIDFactory as MessageFactory
_ = MessageFactory("Default")

class DocumentBrowserWidget(TextWidget):

    displayWidth = 30

    def _toFieldValue(self, input):
        try:
            if not input:
                return input
            input = str(input)
            context = self.context.context
            portal_url = getToolByName(context, 'portal_url')
            portal_path = portal_url.getPortalPath()
            if not input.startswith(portal_path):
                if input[0] != '/':
                    input = '/' + input
                input = portal_path + input
            
            obj = context.unrestrictedTraverse(input)
            return u'/'.join(obj.getPhysicalPath())
        except (AttributeError, KeyError), v:
            raise ConversionError("Selected object not found", v)

    def __call__(self):
        value = self._getFormValue()
        res = renderElement(self.tag,
                            type=self.type,
                            name=self.name,
                            id=self.name,
                            value=value,
                            cssClass=self.cssClass,
                            style=self.style,
                            size=self.displayWidth,
                            extra=self.extra)
        style = "text-decoration:none;border:1px solid black;padding:0.2em 0.3em;"
        js = "window.open('documentnavigation_popup?input_id=" + self.name + \
                          "&current_uid=..', 'DirectoryMultiEntryFinder', " \
                          "'toolbar=0, scrollbars=1, location=0, " \
                          "statusbar=0, menubar=0, resizable=1, dependent=1, " \
                          "width=500, height=480')"
        btn = renderElement("input", type="button", style=style, onClick=js, 
                            value=_("..."))
        return res + "&nbsp;" + btn


form/interfaces.py
# -*- coding: ISO-8859-15 -*-
# (C) Copyright 2005 Nuxeo SARL <http://nuxeo.com>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License version 2 as published
# by the Free Software Foundation.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
# 02111-1307, USA.
#
# $Id: interfaces.py 23798 2005-06-10 16:16:35Z lregebro $

from zope.interface import Interface

class IDocumentNavigation(Interface):
    """Support for popup document navigation windows"""
    
    def getDocumentNavigation(finder, root_uid, current_uid=None, REQUEST=None):
        """Returns a CPSNavigation object

        The navigation popup uses a CPSNavigation object to display a 
        navigation tree."""
        


__init__.py
# -*- coding: ISO-8859-15 -*-
# (C) Copyright 2005 Nuxeo SARL <http://nuxeo.com>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License version 2 as published
# by the Free Software Foundation.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
# 02111-1307, USA.
#
# $Id: zopecal.py 24305 2005-06-22 14:39:11Z lregebro $

""" CPSonFive

This product is the container for any reusable integration between CPS and Five.
"""
from zope.interface import implements

from types import StringTypes
from zope.schema.interfaces import ITitledTokenizedTerm
from zope.i18nmessageid import Message

# Zope3.0.0 doesn't translates dropdownboxes (that's a bug)
def textForValue(self, term):
    """Extract a string from the term.

    The term must be a vocabulary tokenized term.

    This can be overridden to support more complex term objects. The token
    is returned here since it's the only thing known to be a string, or
    str()able."""
    if ITitledTokenizedTerm.providedBy(term):
        return self.translate(term.title)
    return self.translate(term.token)


from zope.i18n.interfaces import IUserPreferredCharsets

class ISO15Charset(object):
    # This object implements the selector function for IUserPreferredCharsets
    # but doesn't care what the user prefer, It returns ISO-8859-15 anyway.
    # XXX For support of non-european languages, we need to match this with 
    # the selected UI language somehow. 
    implements(IUserPreferredCharsets)

    def __init__(self, request):
        self.request = request
        
    def getPreferredCharsets(self):
        '''See interface IUserPreferredCharsets'''
        return ['iso-8859-15']

from zope.event import notify
from Products.Five.form import EditView, Update, applyWidgetsChanges
from zope.app.form.interfaces import WidgetsError
from zope.app.form.utility import setUpEditWidgets, applyWidgetsChanges
from zope.app.event.objectevent import ObjectModifiedEvent
try:
    from zope.app.i18n import ZopeMessageFactory as _
except:
    from zope.app.i18n import ZopeMessageIDFactory as _
import transaction, datetime

def EditViewUpdate(self):
    if self.update_status is not None:
        # We've been called before. Just return the status we previously
        # computed.
        return self.update_status

    status = ''

    content = self.adapted

    if Update in self.request.form.keys():
        changed = False
        try:
            changed = applyWidgetsChanges(self, self.schema,
                target=content, names=self.fieldNames)
            # We should not generate events when an adapter is used.
            # That's the adapter's job.  We need to unwrap the objects to
            # compare them, as they are wrapped differently.
            # Additionally, we can't use Acquisition.aq_base() because
            # it strangely returns different objects for these two even
            # when they are identical.  In particular
            # aq_base(self.adapted) != self.adapted.aq_base :-(
            if changed and getattr(self.context, 'aq_base', self.context)\
                        is getattr(self.adapted, 'aq_base', self.adapted):
                notify(ObjectModifiedEvent(content))
        except WidgetsError, errors:
            self.errors = errors
            status = _("An error occurred.")
            transaction.abort()
        else:
            setUpEditWidgets(self, self.schema, source=self.adapted,
                             ignoreStickyValues=True,
                             names=self.fieldNames)
            if changed:
                self.changed()
                # CPS Patch begins here.
                localizer = getattr(self.context, 'Localizer', None)
                if localizer is not None:                    
                    status = localizer.default("Updated on %(date_time)s")
                    format = str(localizer.default('date_medium'))
                    date = datetime.datetime.now().strftime(format)
                else:
                    date = str(datetime.utcnow())
                    
                status = _("Updated on ${date_time}",
                           mapping={'date_time': date})

    self.update_status = status

    return status

from zope.i18n.interfaces import IUserPreferredLanguages, ILanguageAvailability

class UserLanguages(object):
    """Return the preferred language.
    
    This one cares about the LOCALIZER_LANGUAGE used by CPS/Localizer to
    explicitly select a language. The LocalizerLanguages component in Five
    also does this, but via the overcomplicated Localizer.AcceptLanguage 
    who seldom returns what we want or expect.
    So I use this instead. //Lennart."""
    implements(IUserPreferredLanguages)

    def __init__(self, request):
        self.request = request

    def getPreferredLanguages(self):
        accept_language = self.request.HTTP_ACCEPT_LANGUAGE
        localizer_language = self.request.get('LOCALIZER_LANGUAGE')
        accept_language = '%s;q=2.0,%s' % (localizer_language, accept_language)
        # Normalize: Remove any spaces, change _ to -, make all lowercase:
        accept_language = accept_language.replace(' ', '')
        accept_language = accept_language.replace('_', '-')
        accept_language = accept_language.lower()
        langs = {}
        for lang in accept_language.split(','):
            if lang.find(';') != -1:
                lang, q = lang.split(';')
                q = float(q[2:]) # remove the "q=" and make into a float
            else:
                q = 1.0
            if lang not in langs:
                langs[lang] = q
                    
        # Make into a list of tuples with value first, and key last 
        # for easy sorting:
        langs = [(q,l) for (l,q) in langs.items()]
        langs.sort()
        langs.reverse()
        # Return languages in order of quality
        return [l for q, l in langs]

class CPSLanguages(object):
    """Returns the languages that exist in a CPS site"""
    
    implements(ILanguageAvailability)

    def __init__(self, context):
        self.context = context
        
    def getAvailableLanguages(self):
        pts = self.context.translation_service
        languages = []
        for lang in [pts.getDefaultLanguage()] + pts.getSupportedLanguages():
            # Localizer uses the format xx_YY instead of xx-yy:
            lang = lang.replace('_', '-')
            lang = lang.lower()
            if lang not in languages:
                languages.append(lang)
        return languages        
    
def initialize(context):

    # Zope3monkey
    from zope.app.form.browser.itemswidgets import ItemsWidgetBase
    ItemsWidgetBase.textForValue = textForValue
    # Five monkey
    EditView.update = EditViewUpdate
